### PR TITLE
feat(otp2): add OTP2 support

### LINKED ISSUE #44: Cannot login via CLI - OTP2 not supported
Hello,

I can login fine on the browser.
However the CLI keeps saying master password incorrect after trying to login by running `dcli-win sync`.
`The master password you provided is incorrect, would you like to retry?`

### PR DESCRIPTION
**Pull request by ni0ki**
- Add support for server key at device registration / login
- Prevent the local key from being saved in the keychain if user has OTP2. This will force the cli to ask for the OTP2/Master password to be able to decipher the local key in the DB

This raises another problem, which is that we don't check for revoked devices before logging in. As such, if OTP2 is disabled by the user on another device, we won't be able to see it here, and the authentication will simply fail... But this is for another MR.

(I'm fixing this "as it is" but I guess the architecture could be improved a bit. Let me know if it's OK for you to go on this way for now)

Should fix #44 meanwhile

### CODE REVIEW

**Review by Mikescops — APPROVED** [review]
The changes look good, I'm indeed concerned about having users logging in everytime they request a password.
Protecting the combo serverKey+MP with the keychain is acceptable from a security standpoint, I think it will make the CLI more usable for OTP2 users.

File: src/middleware/perform2FAVerification.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            'Unable to check 2FA Status. This will prevent you from logging in if you are using 2FA at each login (OTP2)'
+        );
+        return;
+    }
+
+    if (twoFactorAuthStatus.type === 'totp_login') {
+        const otp = await askOtp();
+        ({ authTicket } = await performTotpVerification({
+            login,
+            otp: String(otp).padStart(5, '0'), // Buggy if OTP starts with 0
```
[the remark this answers is not in the document]
Comment (Mikescops) [reply src/middleware/perform2FAVerification.ts:29]: otp is already a string with the correct length

**Comment by ni0ki** [comment]
I updated the MR following our discussion:

- Authentication type (well.. 2FA to be precise) is now stored into the Device configuration
- If required, the server key will be integrated into the master password (which should be renamed towards something like sessionKey).
- In case of missing keychain only, then we ask the user for OTP2 + Master password.

This raises the question of if/how should we check for authentication changes or device revocation, as the cli never rechecks against the server.

File: src/database/upgrade/upgrade.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import Database from 'better-sqlite3';
+
+interface RecordCount {
+    CNTREC: number;
+}
+
+export const upgradeDatabase = (db: Database.Database) => {
+    const authenticatioFieldCheck = db
```
Comment (Mikescops) [inline src/database/upgrade/upgrade.ts:8]:
```suggestion
    const authenticationFieldCheck = db
```

**Review by Mikescops — APPROVED** [review]
Sounds good

**Comment by Mikescops** [comment]
> This raises the question of if/how should we check for authentication changes or device revocation, as the cli never rechecks against the server.

As long as we're doing read only I think we're fine. For revocation we could include a check yes and reset the DB.

File: src/endpoints/performTotpVerification.ts [hunk 3]
```diff
@@ -2,7 +2,7 @@ import { requestApi } from '../requestApi';
 
 interface PerformTotpVerification {
     login: string;
-    otp: string;
+    otp: number;
```
Comment (Mikescops) [inline src/endpoints/performTotpVerification.ts:5]: i think it's still a string no?

File: src/middleware/perform2FAVerification.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ni0ki) [reply src/middleware/perform2FAVerification.ts]: Yup sorry, did not update my fork main branch. Should be fixed in the latest commit

File: src/endpoints/performTotpVerification.ts (hunk 3 above)
Comment (ni0ki) [reply src/endpoints/performTotpVerification.ts]: Main branch was not up to date. Should be better now

**Pull request merged**